File: app/src/main/java/starbright/com/projectegg/features/userAccount/UserAccountPresenter.java
package starbright.com.projectegg.features.userAccount;

import android.support.annotation.NonNull;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;

import java.util.regex.Pattern;

import starbright.com.projectegg.util.Constants;

class UserAccountPresenter implements UserAccountContract.Presenter{

    private final UserAccountContract.View mView;
    private final FirebaseAuth mAuth;

    private boolean mIsLoginAuth;
    private FirebaseAuth.AuthStateListener mAuthStateListener =
            new FirebaseAuth.AuthStateListener() {
                @Override
                public void onAuthStateChanged(@NonNull FirebaseAuth firebaseAuth) {
                    if (isUserEmailVerified(firebaseAuth)) {
                        mView.navigateToSearchRecipePage();
                    }
                }
            };

    UserAccountPresenter(UserAccountContract.View view, boolean isLogin) {
        mView = view;
        mView.setPresenter(this);
        mIsLoginAuth = isLogin;
        mAuth = FirebaseAuth.getInstance();
    }

    @Override
    public void handleNavigationTextButtonClicked() {
        mIsLoginAuth = !mIsLoginAuth;
        mView.updateView(mIsLoginAuth);
    }

    @Override
    public void handleAuthenticationButtonClicked(String email, String password) {
        if (email == null || email.isEmpty()) {
            if (validateEmailFormat(email)) {
                mView.showEmailFormatWrongErrorToast();
            } else {
                mView.showEmailEmptyErrorToast();
            }
            return;
        } else if (password == null || password.isEmpty()){
            mView.showPasswordErrorToast();
            return;
        }

        mView.showProgressBar();
        mView.disableView();
        final Task<AuthResult> auth = isLoginAuthentication() ? login(email, password)
                : signup(email, password);
        auth
                .addOnSuccessListener(new OnSuccessListener<AuthResult>() {
                @Override
                public void onSuccess(AuthResult authResult) {
                    if (!isLoginAuthentication()) {
                        sendUserVerificationEmail(authResult.getUser());
                    } else {
                        mView.hideProgressBar();
                        mView.enableView();
                        if (!authResult.getUser().isEmailVerified()) {
                            mView.showVerificationEmailSentDialog();
                        }
                    }
                }
            })
            .addOnFailureListener(new OnFailureListener() {
                @Override
                public void onFailure(@NonNull Exception e) {
                    mView.hideProgressBar();
                    mView.enableView();
                    mView.showLoginErrorToast(e.getMessage());
                }
            });
    }

    @Override
    public void onDestroy() {
        mAuth.removeAuthStateListener(mAuthStateListener);
    }

    @Override
    public void start() {
        mAuth.addAuthStateListener(mAuthStateListener);
       mView.setupProgressBar();
        mView.updateView(mIsLoginAuth);
    }

    @Override
    public boolean isLoginAuthentication() {
        return mIsLoginAuth;
    }

    private Task<AuthResult> login(String email, String password) {
        return mAuth.signInWithEmailAndPassword(email, password);
    }

    private Task<AuthResult> signup(String email, String password) {
        return mAuth.createUserWithEmailAndPassword(email, password);
    }

    private void sendUserVerificationEmail(FirebaseUser user) {
        user.sendEmailVerification()
                .addOnSuccessListener(new OnSuccessListener<Void>() {
                    @Override
                    public void onSuccess(Void aVoid) {
                        mView.hideProgressBar();
                        mView.enableView();
                        mView.showVerificationEmailSentDialog();
                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        mView.hideProgressBar();
                        mView.enableView();
                        mView.showLoginErrorToast(e.getMessage());
                    }
                });
    }

    private boolean isUserEmailVerified(FirebaseAuth firebaseAuth) {
        if (firebaseAuth.getCurrentUser() != null) {
            return firebaseAuth.getCurrentUser().isEmailVerified();
        } else {
            return false;
        }
    }

    private boolean validateEmailFormat(String email) {
        return Pattern.compile(Constants.EMAIL_FORMAT).matcher(email).matches();
    }
}

File: app/src/main/java/starbright/com/projectegg/features/userAccount/UserAccountPresenter.kt
/**
 * Created by Andreas on 7/10/2019.
 */

package starbright.com.projectegg.features.userAccount

class UserAccountPresenter(
        private val mView: UserAccountContract.View,
        isLogin: Boolean
) : UserAccountContract.Presenter {
    override fun handleAuthenticationButtonClicked(email: String, password: String) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun handleOnBackPressedClicked(): Boolean {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun updateUserAuthRole() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun onDestroy() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun handleVerificationEmailDialogClicked() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

//    private val mAuth: FirebaseAuth
//
//    private var isLoginAuthentication: Boolean = false
//
//    private val mAuthStateListener = FirebaseAuth.AuthStateListener { firebaseAuth ->
//        if (isUserEmailVerified(firebaseAuth)) {
//            mView.navigateToSearchRecipePage()
//        }
//    }
//
//    init {
//        mView.setPresenter(this)
//        isLoginAuthentication = isLogin
//        mAuth = FirebaseAuth.getInstance()
//    }
//
//    override fun updateUserAuthRole() {
//        isLoginAuthentication = !isLoginAuthentication
//        mView.updateView(isLoginAuthentication)
//    }
//
//    override fun handleVerificationEmailDialogClicked() {
//        if (!isLoginAuthentication) {
//            updateUserAuthRole()
//        }
//    }
//
//    override fun handleAuthenticationButtonClicked(email: String, password: String) {
//        when {
//            email.isEmpty() -> {
//                mView.showEmailEmptyErrorToast()
//                return
//            }
//            validateEmailFormat(email) -> {
//                mView.showEmailFormatWrongErrorToast()
//                return
//            }
//            password.isEmpty() -> {
//                mView.showPasswordErrorToast()
//                return
//            }
//        }
//
//        mView.showProgressBar()
//        mView.disableView()
//        val authSubscriber = if (isLoginAuthentication)
//            login(email, password)
//        else
//            signup(email, password)
//
//        authSubscriber
//                .addOnSuccessListener { authResult ->
//                    if (!isLoginAuthentication) {
//                        sendUserVerificationEmail(authResult.user)
//                    } else {
//                        mView.hideProgressBar()
//                        mView.enableView()
//                        if (!authResult.user.isEmailVerified) {
//                            mView.showVerificationEmailSentDialog()
//                        }
//                    }
//                }
//                .addOnFailureListener { e ->
//                    mView.hideProgressBar()
//                    mView.enableView()
//                    mView.showLoginErrorToast(e.message)
//                }
//    }
//
//    override fun handleOnBackPressedClicked(): Boolean {
//        return if (!isLoginAuthentication) {
//            updateUserAuthRole()
//            true
//        } else {
//            false
//        }
//    }
//
//    override fun onDestroy() {
//        mAuth.removeAuthStateListener(mAuthStateListener)
//    }
//
//    override fun start() {
//        mAuth.addAuthStateListener(mAuthStateListener)
//        mView.setupProgressBar()
//        mView.updateView(isLoginAuthentication)
//    }
//
//    private fun login(email: String, password: String): Task<AuthResult> {
//        return mAuth.signInWithEmailAndPassword(email, password)
//    }
//
//    private fun signup(email: String, password: String): Task<AuthResult> {
//        return mAuth.createUserWithEmailAndPassword(email, password)
//    }
//
//    private fun sendUserVerificationEmail(user: FirebaseUser) {
//        user.sendEmailVerification()
//                .addOnSuccessListener {
//                    mView.hideProgressBar()
//                    mView.enableView()
//                    mView.showVerificationEmailSentDialog()
//                }
//                .addOnFailureListener { e ->
//                    mView.hideProgressBar()
//                    mView.enableView()
//                    mView.showLoginErrorToast(e.message)
//                }
//    }
//
//    private fun isUserEmailVerified(firebaseAuth: FirebaseAuth): Boolean {
//        return firebaseAuth.currentUser?.isEmailVerified ?: false
//    }
//
//    private fun validateEmailFormat(email: String): Boolean {
//        return Pattern.compile(Constants.EMAIL_FORMAT).matcher(email).matches()
//    }
}
